lib/lz4.c

```diff
@@ -429,22 +429,42 @@ LZ4_FORCE_O2_INLINE_GCC_PPC64LE void
 LZ4_memcpy_using_offset(BYTE* dstPtr, const BYTE* srcPtr, BYTE* dstEnd, const size_t offset)
 {
     BYTE v[8];
+    U16 u16;
+    U32 u32;
+    U64 u64;
 
     assert(dstEnd >= dstPtr + MINMATCH);
     LZ4_write32(dstPtr, 0);   /* silence an msan warning when offset==0 */
 
     switch(offset) {
     case 1:
-        memset(v, *srcPtr, 8);
+        if(sizeof(void*) == 8) {
```

Comment:
I guess this change will hurt ARM64 performance.

Reply:
Can you elaborate?

Reply:
Based on my previous experience on ARM64 LZ4 performance tunning. I realized that if-else will affect the performance obviously.
If we can confirm whether (sizeof(void*) == 8) during compiling, that would be better than runtime I guess.

Reply:
This is a compile-time check, even at `-O0`.

Reply:
hmm...
Which compiler did you use? Clang or GCC?

Reply:
I will check on ARM64.

Reply:
The objcode show the similar code.
            u64 = *srcPtr * 0x0101010101010101;
            memcpy(v, &u64, 8);
Vanilla
        memset(v, *srcPtr, 8);
    a560:	39400030 	ldrb	w16, [x1]
    a564:	1b0d7e10 	mul	w16, w16, w13

Patched:
            u64 = *srcPtr * 0x0101010101010101;
    a550:	39400031 	ldrb	w17, [x1]
    a554:	9b0e7e31 	mul	x17, x17, x14
The objdump is same.

            memcpy(&u16, srcPtr, 2);
            u64 = u16 * 0x0001000100010001;
            memcpy(v, &u64, 8);
Vanilla:
        memcpy(v, srcPtr, 2);
    a58c:	79400030 	ldrh	w16, [x1]
        memcpy(&v[2], srcPtr, 2);
    a590:	33103e10 	bfi	w16, w16, #16, #16
Patched:
            u64 = u16 * 0x0001000100010001;
    a580:	79400031 	ldrh	w17, [x1]
    a584:	9b0d7e31 	mul	x17, x17, x13

            memcpy(&u32, srcPtr, 4);
            u64 = u32 | (((U64)u32) << 32);
            memcpy(v, &u64, 8);
Vanilla:
        memcpy(v, srcPtr, 4);
    b6c8:	b94000a5 	ldr	w5, [x5]
    memcpy(dstPtr, v, 8);
    b6d0:	29001585 	stp	w5, w5, [x12]
Patched:
            u64 = u32 | (((U64)u32) << 32);
    d9a4:	b94001ce 	ldr	w14, [x14]
    d9a8:	b3607dce 	bfi	x14, x14, #32, #32

Benchmark result will be updated later.

Reply:
ARM64:
Vanilla:
violet:/data # ./lzbench   -elz4 -t4,60 /sdcard/silesia.tar                                                                                                                                                
lzbench 1.7.3 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                   6011 MB/s  5833 MB/s   211957760 100.00 /sdcard/silesia.tar
lz4 1.9.x                 322 MB/s  1915 MB/s   100881381  47.60 /sdcard/silesia.tar

Patched:
violet:/data # ./lz_m  -elz4 -t4,60 /sdcard/silesia.tar                                                                                                                                                    
lzbench 1.7.3 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                   6109 MB/s  5876 MB/s   211957760 100.00 /sdcard/silesia.tar
lz4 1.9.x                 322 MB/s  1917 MB/s   100881381  47.60 /sdcard/silesia.tar


Same level on ARM64.

Reply:
This is indeed a static test, solved at compilation time.

@wolfpld , I believe that what you want to test is if the target architecture supports 64-bit registers.
Such a test is slightly different from `sizeof(void*)`, because some environments use 32-bit addresses while still enjoying the full 64-bit capabilities of the cpu. Yeah, that seems weird, but I can think of 2 : Linux' `x32`, and `wasm` (this last one I was told, I did not checked directly). There might be others.

To fix these cases, `lz4` supports a different test : `sizeof(reg_t)==8`.
The magic is in the definition of `reg_t`, which can be perfectible, but at least, we have one place to concentrate efforts and debugging if the detection fail for one specific platform.

That's basically my only comment for this PR.
I will also try to measure speed, so my understanding is that one should not expect much variation.

Reply:
Good catch, fixed.